File: Timing/timing.js
var fs = require('fs');
var clock = require('posix-clock');
var NanoTimer = require('nanotimer');
var nanoTimerObject = new NanoTimer();

jdata {
    char* logTime as logger;
    char* broadcastTime as broadcaster;
}

var numTests = 1000;
var clockResolution = clock.getres(clock.MONOTONIC);
var results = new Array();
var tests = {
	seqAsync: new Array(),
	parAsync: new Array(),
	sync: new Array(),
	logger: new Array(),
	jSyncRound: new Array(),
	jAsyncRound: new Array()
}




jasync function seqAsync(startSec, startNS) {

	var clockTime = clock.gettime(clock.MONOTONIC);
	var secDelay = clockTime.sec - startSec;
	var nsDelay = clockTime.nsec - startNS;


	if(secDelay > 0) {
		nsDelay = 1000000000 - nsDelay;
	}

	tests.seqAsync.push(nsDelay);
    asyncDone();
}

jasync function parAsync(startSec, startNS) {
	var clockTime = clock.gettime(clock.MONOTONIC);
	var secDelay = clockTime.sec - startSec;
	var nsDelay =  clockTime.nsec - startNS;
	if(secDelay > 0) {
		nsDelay = 1000000000 - nsDelay;
	}
	tests.parAsync.push(nsDelay);
	if(tests.parAsync.length === numTests) {
		asyncDone2();
	}
}

jsync function syncJS(startSec, startNS) {
	var clockTime = clock.gettime(clock.MONOTONIC);
	var secDelay = clockTime.sec - startSec;
	var nsDelay =  clockTime.nsec - startNS;
	if(secDelay > 0) {
		nsDelay = 1000000000 - nsDelay;
	}
	tests.sync.push(nsDelay);
	return 0;
}

jsync function emptyJS() {
	return 0;
}

jsync function startLoggerTest() {
	var counter = 0;
	nanoTimerObject.setInterval(function() {
	    if (logTime[1] !== undefined) {
    		if(logTime[1].index_of_last_value > counter) {
	    		var clockTime = clock.gettime(clock.MONOTONIC);
	    		var timeString = clockTime.sec + clockTime.nsec.toString().padStart(9, "0");
	    		var delay = Number(timeString) - Number(logTime[1].lastValue());
	    		tests.logger.push(delay);
	    		counter = logTime[1].index_of_last_value;
	    	}
    		if(counter === numTests) {
    			nanoTimerObject.clearInterval();
    		}
	    }
	}, '', '1u');
	return 0;
}


function jSyncRound() {
	for (var i = 0; i < numTests; i++) {
		var startTime = clock.gettime(clock.MONOTONIC);
		cSyncReturn();

		var endTime = clock.gettime(clock.MONOTONIC);
		var secDelay = endTime.sec - startTime.sec;
		var nsDelay =  endTime.nsec - startTime.nsec;
		if(secDelay > 0) {
			nsDelay = 1000000000 - nsDelay;
		}

		tests.jSyncRound.push(nsDelay);
	}
}

function jAsyncRound() {
	for (var i = 0; i < numTests; i++) {
		var startTime = clock.gettime(clock.MONOTONIC);

		cAsyncReturn();

		var endTime = clock.gettime(clock.MONOTONIC);
		var secDelay = endTime.sec - startTime.sec;
		var nsDelay =  endTime.nsec - startTime.nsec;
		if(secDelay > 0) {
			nsDelay = 1000000000 - nsDelay;
		}

		tests.jAsyncRound.push(nsDelay);
	}
}

function broadcastTest() {
	var counter = 0;
	var interval = setInterval(function() {
		var clockTime = clock.gettime(clock.MONOTONIC);
	    var timeString = clockTime.sec + clockTime.nsec.toString().padStart(9, "0");
		broadcastTime.broadcast(timeString);
		counter++;
		// if(counter === numTests) {
		// 	clearInterval(interval);
		// }
	}, 10);
}
// jasync function startLoggerTest() {
// 	counter = 0;
// 	while(logTime[1] === undefined) {
// 		console.log(logTime[1]);
// 		console.log("HEY");
// 	}
// 	while(true) {
// 		console.log(logTime[1].lastValue());
// 		if(logTime[1].lastValue() == counter) {
// 		    counter++;
// 		    console.log(counter);
// 		    if(counter === numTests) {
// 		        break;
// 		    }
// 		}
// 	}
// }
jasync function writeResults() {
	console.log("Writing Results");
	fs.writeFile('seqAsync.txt', tests.seqAsync.join("\n"), (err) => {
	  if (err) throw err;
	});
	// fs.writeFile('parAsync.txt', tests.parAsync.join("\n"), (err) => {
	//   if (err) throw err;
	// });
	// fs.writeFile('sync.txt', tests.sync.join("\n"), (err) => {
	//   if (err) throw err;
	// });
	// fs.writeFile('logger.txt', tests.logger.join("\n"), (err) => {
	//   if (err) throw err;
	// });
	// fs.writeFile('jSyncRound.txt', tests.jSyncRound.join("\n"), (err) => {
	//   if (err) throw err;
	// });
	// fs.writeFile('jAsyncRound.txt', tests.jAsyncRounds.join("\n"), (err) => {
	//   if (err) throw err;
	// });
}

// broadcastTest();

File: Timing/timing.c
#include <time.h>
int numTests = 1000;
int iteration = 0;
// int results[100];
struct timespec tp;
clockid_t clk_id = CLOCK_MONOTONIC;

void seqAsync(int, int);
void parAsync(int, int);
int syncJS(int, int);
int syncJSCond(int, int);
void writeResults();
int emptyJS();
void startLoggerTest();
void syncRoundTrip();
void syncCallTime();
void loggerTime();
void asyncCallTimeParallel();
void broadcastTest();
void syncCondCallTime();

jsync int cSyncReturn() {
	return 0;
}

jasync cAsyncReturn() {
	return;
}

void asyncCallTimeSequencial() {
	clock_gettime(clk_id, &tp);
	seqAsync(tp.tv_sec, tp.tv_nsec);
}


jasync asyncDone() {
	iteration++;
	if(iteration < numTests) {
		asyncCallTimeSequencial();
	} else {
		asyncCallTimeParallel();
	}
}


void asyncCallTimeParallel() {
	for (int i = 0; i < numTests; i++) {
		clock_gettime(clk_id, &tp);
		parAsync(tp.tv_sec, tp.tv_nsec);
	}
}

jasync asyncDone2() {
	syncCallTime();
}


void syncCallTime() {
	for (int i = 0; i < numTests; i++) {
		clock_gettime(clk_id, &tp);
		syncJS(tp.tv_sec, tp.tv_nsec);
	}
	syncCondCallTime();
}

void syncCondCallTime() {
	for (int i = 0; i < numTests; i++) {
		clock_gettime(clk_id, &tp);
		syncJSCond(tp.tv_sec, tp.tv_nsec);
	}
	syncRoundTrip();
}

void syncRoundTrip() {
	struct timespec start, stop;
	double accum;
	int delay;
	FILE *f = fopen("syncRound.txt", "w");
	for (int i = 0; i < numTests; i++) {
		clock_gettime(clk_id, &start);

		emptyJS();

		clock_gettime(clk_id, &stop);
		delay = stop.tv_nsec - start.tv_nsec;
		// accum = ( stop.tv_sec - start.tv_sec )
	 //          + ( stop.tv_nsec - start.tv_nsec )
	 //            / 1000000000L;
	    // fprintf(f, "%f\n", accum);
	    fprintf(f, "%i\n", delay);
	}
	fclose(f);
	loggerTime();
}

void loggerTime() {
	struct timespec sleepValue;
   	sleepValue.tv_sec = 0;
   	sleepValue.tv_nsec = 100000000L;
	startLoggerTest();
	char buf[50];
	for (int i = 0; i < numTests; i++) {
		clock_gettime(clk_id, &tp);
		snprintf(buf, 50, "%lu%09lu", tp.tv_sec, tp.tv_nsec);
		logTime = buf;
	//	usleep(100);
	}

}

jasync broadcastStarter() {
	int count = 0;
	char* lastVal;
	char* curVal;
	char buf[50];
	char *ptr;
	broadcastTest();

	FILE *f = fopen("broadcast.txt", "w");
	while(1) {
		curVal = broadcastTime;
		if(curVal != lastVal) {
			clock_gettime(clk_id, &tp);
			snprintf(buf, 50, "%lu%09lu", tp.tv_sec, tp.tv_nsec);

	   		fprintf(f, "%lu\n", strtol(buf, &ptr, 10) - strtol(curVal, &ptr, 10));

			lastVal = curVal;
			count++;
			if(count == numTests) {
				break;
			}
		}
	}
	fclose(f);

	writeResults();
}


void main() {

	asyncCallTimeSequencial();
}
